services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/switches/SwitchRulesSpec.groovy

```diff
@@ -114,6 +109,86 @@ class SwitchRulesSpec extends BaseSpecification {
+        def offset = sw.defaultCookies.contains(Cookie.CATCH_BFD_RULE_COOKIE) ? 0 : 1
+        installedRules.size() == data.rulesInstalled(defaultRules.size() + offset)
+        Wrappers.wait(RULES_INSTALLATION_TIME) {
+            compareRules(northbound.getSwitchRules(sw.dpId).flowEntries, data.expectedRules(defaultRules))
+        }
+
+        and: "Install missing default rules"
+        Wrappers.wait(RULES_INSTALLATION_TIME) {
+            northbound.installSwitchRules(sw.dpId, InstallRulesAction.INSTALL_DEFAULTS)
+            assert northbound.getSwitchRules(sw.dpId).flowEntries.size() == defaultRules.size()
+        }
+
+        where:
+        [data, sw] << [
+                [
+                        [
+                                installRulesAction: InstallRulesAction.INSTALL_DROP,
+                                rulesInstalled    : { rulesCount -> 1 },
+                                expectedRules     : { rules ->
+                                    rules.findAll {
+                                        it.cookie == Cookie.DROP_RULE_COOKIE
+                                    }
+                                }
+                        ],
+                        [
+                                installRulesAction: InstallRulesAction.INSTALL_BROADCAST,
+                                rulesInstalled    : { rulesCount -> 1 },
+                                expectedRules     : { rules ->
+                                    rules.findAll {
+                                        it.cookie == Cookie.VERIFICATION_BROADCAST_RULE_COOKIE
+                                    }
+                                }
+                        ],
+                        [
+                                installRulesAction: InstallRulesAction.INSTALL_UNICAST,
+                                rulesInstalled    : { rulesCount -> 1 },
+                                expectedRules     : { rules ->
+                                    rules.findAll {
+                                        it.cookie == Cookie.VERIFICATION_UNICAST_RULE_COOKIE
+                                    }
+                                }
+                        ],
+                        [
+                                installRulesAction: InstallRulesAction.INSTALL_BFD_CATCH,
+                                rulesInstalled    : { rulesCount -> 1 },
+                                expectedRules     : { rules ->
+                                    rules.findAll {
+                                        it.cookie == Cookie.CATCH_BFD_RULE_COOKIE
+                                    }
+                                }
+                        ],
+                        [
+                                installRulesAction: InstallRulesAction.INSTALL_DEFAULTS,
+                                rulesInstalled    : { rulesCount -> rulesCount },
+                                expectedRules     : { rules -> rules }
```

Comment:
This iteration brings a lot of complication to the test. If you exclude this iteration you could drop both closures (expectedRules would be just a Cookie.SOMETHING, rulesInstalled could be removed completely). This will simplify the 'where' block and the test code overall. Plus when testing the `INSTALL_DEFAULTS` you use the same `INSTALL_DEFAULTS` in the cleanup which is excess. I think it's better to duplicate some simple code and have a separate test for `INSTALL_DEFAULTS` over making a single test which is harder to understand.
Though it can still work as it is, I don't insist

Reply:
Fixed, thanks.